include/kframework/lemmas/bytes-simplification.k

```diff
@@ -68,50 +68,155 @@ module BYTES-SIMPLIFICATION [symbolic]
 
     // #range
 
-    rule #range(_, _, K) => .Bytes requires K <=Int 0 [simplification]
+    rule [range-empty]:
+      #range(_:Bytes, S:Int, W:Int) => .Bytes
+      requires S <Int 0 orBool W <=Int 0
+      [simplification(1)]
+
+    rule [range-full]:
+      #range(B:Bytes, 0, W:Int) => B
+      requires lengthBytes(B) ==Int W
+      [simplification]
+
+    rule [range-overfull]:
+      #range(B:Bytes, 0, W:Int) => B +Bytes #buf(W -Int lengthBytes(B), 0)
+      requires lengthBytes(B) <Int W
+      [simplification]
+
+    rule [range-params-equal-ml]:
+      { #range (B1:Bytes, S1:Int, W1:Int) #Equals #range (B2:Bytes, S2:Int, W2:Int) } => #Top
+      requires B1 ==K B2 andBool S1 ==Int S2 andBool W1 ==Int W2
+      [simplification]
+
+    rule [range-included-in-cHead]:
+      #range(B1:Bytes +Bytes _:Bytes, S:Int, W:Int) => #range(B1, S, W)
+      requires S +Int W <=Int lengthBytes(B1)
+      [simplification]
 
-    rule #range(#writeRange(M, N, BUF:Bytes ), L, K) =>
-         #let W = minInt(K, N -Int L) #in #range(M, L, W) +Bytes #range(#writeRange(M, N, BUF), N, K -Int W)
-      requires 0 <Int K andBool L <Int N [simplification]
+    rule [range-outside-cHead]:
+      #range(B1:Bytes +Bytes B2:Bytes, S:Int, W:Int) => #range(B2, S -Int lengthBytes(B1), W)
+      requires lengthBytes(B1) <=Int S
+      [simplification]
 
-    rule #range(#writeRange(M, N, BUF:Bytes), L, K) =>
-           #let S = L -Int N #in
-           #let W = minInt(K, lengthBytes(BUF) -Int S) #in
-           #range(BUF, S, W) +Bytes #range(M, N +Int lengthBytes(BUF), K -Int W)
-      requires 0 <Int K andBool N <=Int L andBool L <=Int N +Int lengthBytes(BUF) [simplification]
+    rule [range-includes-cHead]:
+      #range(B1:Bytes +Bytes B2:Bytes, 0, W:Int) => B1 +Bytes #range(B2, 0, W -Int lengthBytes(B1))
+      requires lengthBytes(B1) <=Int W
+      [simplification]
 
-    rule #range(#writeRange(M, N, BUF:Bytes ), L, K) => #range(M, L, K)
-      requires 0 <Int K andBool N +Int lengthBytes(BUF) <=Int L [simplification]
+    rule [range-inside-cHead-base-conc-left]:
+      #range(B:Bytes, S1:Int +Int S2:Int, W:Int) => #range(#range(B, S1, lengthBytes(B) -Int S1), S2, W)
+      requires 0 <Int S1 +Int S2 andBool S1 +Int S2 <=Int lengthBytes(B)
+      [concrete(B), concrete(S1), simplification]
```

Comment:
- `B:Bytes == #buf(6, 1)`
- `S1 == -3`
- `S2 == 7`
- `W == 2`

Then

- `#range(B, S1 +Int S2, W) => #range(B, 4, 2) => #buf(2, 1)`
- `#range(B, S1 +Int S2, W) => #range(#range(B, -3, lengthBytes(B) -Int -3), 7, 2) => #range(.Bytes, 7, 2) => #buf(2, 0)`

Reply:
This same counterexample may apply to other range lemmas you have about. Probably easily solved by requiring that `S1` and `S2` are individually non-negative too.